### PR TITLE
Migrate Session class from Java to Kotlin (part 3)

### PR DESCRIPTION
**Pull request by johnjohndoe**
# Description
- Remove unneeded modification of `Session#hasAlarm` when adding or deleting session alarms.
- Prepare `LayoutCalculator` to continue working with immutable `Session` class.
- Prepare `Session#cancel` to continue working with immutable `Session` class.
- Prepare `StarredListFragment#deleteSession` to continue working with immutable `Session` class.

# Successfully tested on
with `ccc37c3` flavor, `debug` build
- :heavy_check_mark: Samsung Galaxy Tab S7 FE, SM-T733 (tablet), Android 14 (API 34)

# Related
- https://github.com/EventFahrplan/EventFahrplan/issues/631
- https://github.com/EventFahrplan/EventFahrplan/pull/632
- https://github.com/EventFahrplan/EventFahrplan/pull/633

### CODE REVIEW

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/LayoutCalculator.kt [hunk 1]
```diff
@@ -31,17 +30,15 @@ data class LayoutCalculator @JvmOverloads constructor(
     }
 
     fun calculateLayoutParams(roomData: RoomData, conference: Conference): Map</* sessionId */ String, LinearLayout.LayoutParams> {
-        val sessions = roomData.sessions
+        val sessions = roomData.sessions.toMutableList()
```
Comment (cketti) [inline app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/LayoutCalculator.kt:33]: Using a mutable list and constantly reading from that list makes it harder than necessary to follow what is going on.

Getting rid of the mutable list could look like this:
```kotlin
    fun calculateLayoutParams(roomData: RoomData, conference: Conference): Map<SessionId, LinearLayout.LayoutParams> {
        val sessions = roomData.sessions
        var previousSessionEndsAt: Int = conference.firstSessionStartsAt.minuteOfDay
        var startTime: Int
        var margin: Int
        var previousSession: Session? = null
        val layoutParamsBySession = mutableMapOf<SessionId, LinearLayout.LayoutParams>()

        for ((index, session) in sessions.withIndex()) {
            startTime = getStartTime(session, previousSessionEndsAt)

            if (startTime > previousSessionEndsAt) {
                // consecutive session
                margin = calculateDisplayDistance(startTime - previousSessionEndsAt)
                if (previousSession != null) {
                    layoutParamsBySession[previousSession.sessionId]!!.bottomMargin = margin
                    margin = 0
                }
            } else {
                // first session
                margin = 0
            }

            val adjustedSession = if (index < sessions.lastIndex) {
                val nextSession = sessions[index + 1]
                fixOverlappingSessions(session, nextSession)
            } else {
                session
            }

            val sessionId = adjustedSession.sessionId
            if (!layoutParamsBySession.containsKey(sessionId)) {
                layoutParamsBySession[sessionId] = createLayoutParams(adjustedSession)
            }

            layoutParamsBySession[sessionId]!!.topMargin = margin
            previousSessionEndsAt = startTime + adjustedSession.duration
            previousSession = adjustedSession
        }

        return layoutParamsBySession
    }

    @VisibleForTesting
    fun fixOverlappingSessions(session: Session, next: Session): Session {
        return if (next.dateUTC > 0 && next.startsAt.isBefore(session.endsAt)) {
            logging.d(LOG_TAG, """Collision: "${session.title}" + "${next.title}"""")
            // cut current at the end, to match next sessions start time
            val newDuration = session.startsAt.minutesUntil(next.startsAt).toInt()
            Session(session).apply { duration = newDuration }
        } else {
            session
        }
    }
```

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/favorites/StarredListFragment.kt [hunk 2]
```diff
@@ -293,8 +293,8 @@ class StarredListFragment :
     }
 
     private fun deleteSession(session: Session) {
```
Comment (cketti) [inline app/src/main/java/nerd/tuxmobil/fahrplan/congress/favorites/StarredListFragment.kt:295]: Side note: This doesn't appear to actually delete a session. So maybe the method should be renamed. Probably best done in a separate PR.

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/favorites/StarredListFragment.kt [hunk 3]
```diff
@@ -293,8 +293,8 @@ class StarredListFragment :
...
-        session.highlight = false
-        viewModel.delete(session)
+        val toBeDeleted = Session(session).apply { highlight = false }
+        viewModel.delete(toBeDeleted)
```
Comment (cketti) [inline app/src/main/java/nerd/tuxmobil/fahrplan/congress/favorites/StarredListFragment.kt:297]: Instead of creating a short-lived Session instance this could be `viewModel.removeHighlight(session)`. And then the view model could call `repository.deleteHighlight(session.sessionId)` (that has to be written).

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmServicesTest.kt [hunk 4]
```diff
@@ -107,7 +107,7 @@ class AlarmServicesTest {
         // AlarmServices invokes discardSessionAlarm() which is tested separately.
         verifyNoInteractions(formattingDelegate)
         verifyInvokedOnce(repository).deleteAlarmForSessionId("S3")
-        assertThat(session.hasAlarm).isFalse()
+        assertThat(session.hasAlarm).isTrue()
```
[the remark this answers is not in the document]
Comment (johnjohndoe) [reply app/src/test/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmServicesTest.kt]: :heavy_check_mark: Agreed. I also thought about it.

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmServicesTest.kt [hunk 5]
```diff
@@ -121,7 +121,7 @@ class AlarmServicesTest {
         alarmServices.deleteSessionAlarm(session)
         verifyNoInteractions(pendingIntentDelegate)
         verifyNoInteractions(formattingDelegate)
-        assertThat(session.hasAlarm).isFalse()
+        assertThat(session.hasAlarm).isTrue()
```
[the remark this answers is not in the document]
Comment (johnjohndoe) [reply app/src/test/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmServicesTest.kt]: :heavy_check_mark: Removed.

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmServicesTest.kt [hunk 6]
```diff
@@ -65,7 +65,7 @@ class AlarmServicesTest {
 
         alarmServices.addSessionAlarm(session, alarmTimesValues.indexOf("60"))
         // AlarmServices invokes scheduleSessionAlarm() which is tested separately.
-        assertThat(session.hasAlarm).isTrue()
+        assertThat(session.hasAlarm).isFalse()
```
[the remark this answers is not in the document]
Comment (johnjohndoe) [reply app/src/test/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmServicesTest.kt]: :heavy_check_mark: Removed.

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/LayoutCalculator.kt [hunk 7]
```diff
@@ -30,13 +30,13 @@ data class LayoutCalculator @JvmOverloads constructor(
         return standardHeight * minutes / DIVISOR
     }
 
-    fun calculateLayoutParams(roomData: RoomData, conference: Conference): Map<Session, LinearLayout.LayoutParams> {
+    fun calculateLayoutParams(roomData: RoomData, conference: Conference): Map</* sessionId */ String, LinearLayout.LayoutParams> {
```
[the remark this answers is not in the document]
Comment (johnjohndoe) [reply app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/LayoutCalculator.kt]: :heavy_check_mark: Applied.

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/LayoutCalculator.kt (hunk 1 above)
Comment (johnjohndoe) [reply app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/LayoutCalculator.kt]: :heavy_check_mark: Thanks a lot for the cleanup. Cute `.withIndex()` - I never used this before.

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/favorites/StarredListFragment.kt (hunk 2 above)
Comment (johnjohndoe) [reply app/src/main/java/nerd/tuxmobil/fahrplan/congress/favorites/StarredListFragment.kt:295]: :x: I leave the naming in the fragment (UI). In the scope of the screen deleting a session makes sense.

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/favorites/StarredListFragment.kt (hunk 3 above)
Comment (johnjohndoe) [reply app/src/main/java/nerd/tuxmobil/fahrplan/congress/favorites/StarredListFragment.kt]: :heavy_check_mark: Refactored. I like it much better now, too.

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/favorites/StarredListFragment.kt (hunk 2 above)
Comment (cketti) [reply app/src/main/java/nerd/tuxmobil/fahrplan/congress/favorites/StarredListFragment.kt:295]: You're removing a session from the list in that screen. That's not the same thing as deleting a session.

Comment (johnjohndoe) [reply app/src/main/java/nerd/tuxmobil/fahrplan/congress/favorites/StarredListFragment.kt]: :heavy_check_mark: Function inlined.

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmServicesTest.kt [hunk 8]
```diff
@@ -107,21 +106,19 @@ class AlarmServicesTest {
...
 
     @Test
-    fun `deleteSessionAlarm resets the session alarm flag when no alarms was scheduled`() {
+    fun `deleteSessionAlarm returns without action when no alarm was scheduled`() {
         val pendingIntentDelegate = mock<PendingIntentDelegate>()
         val formattingDelegate = mock<FormattingDelegate>()
         val alarmServices = createAlarmServices(pendingIntentDelegate = pendingIntentDelegate, formattingDelegate = formattingDelegate)
         whenever(repository.readAlarms(any())) doReturn emptyList()
-        val session = Session("S4").apply { hasAlarm = true }
+        val session = Session("S4") // hasAlarm is not relevant here
```
Comment (cketti) [inline app/src/test/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmServicesTest.kt:117]: It feels like this comment will be more confusing than helpful when the code is read outside the context of the diff in this pull request. Why mention it in a comment if it's not relevant?

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmServicesTest.kt [hunk 9]
```diff
@@ -57,15 +57,14 @@ class AlarmServicesTest {
         whenever(formattingDelegate.getFormattedDateTimeShort(any(), any(), any())) doReturn "not relevant"
         val session = Session("S1").apply {
             dayIndex = 1
-            hasAlarm = false
+            hasAlarm = false // not relevant
```
Comment (cketti) [inline app/src/test/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmServicesTest.kt:60]: Just remove it if it's not relevant.

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/models/SessionTest.kt [hunk 10]
```diff
@@ -301,7 +301,9 @@ class SessionTest {
             changedIsNew = false
             changedIsCanceled = true
         }
-        assertThat(session.apply { cancel() }).isEqualTo(canceledSession)
+        val canceledSession = session.cancel()
+        assertThat(canceledSession).isEqualTo(comparableCanceledSession)
+        assertThat(session).isNotSameInstanceAs(canceledSession)
```
Comment (cketti) [inline app/src/test/java/nerd/tuxmobil/fahrplan/congress/models/SessionTest.kt:306]: This should be the other way round.

```suggestion
        assertThat(canceledSession).isNotSameInstanceAs(session)
```

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/schedule/LayoutCalculatorTest.kt [hunk 11]
```diff
@@ -203,11 +203,12 @@ class LayoutCalculatorTest {
 
         val session1 = createSession(date = conferenceDate, startTime = startTime1, duration = duration1)
         val session2 = createSession(date = conferenceDate, startTime = startTime2, duration = duration2)
-        val sessions = listOf(session1, session2)
+        val sessions = mutableListOf(session1, session2)
```
Comment (cketti) [inline app/src/test/java/nerd/tuxmobil/fahrplan/congress/schedule/LayoutCalculatorTest.kt:206]: There's no need for a list now. Use `session1` and `session2` directly.

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/schedule/LayoutCalculatorTest.kt [hunk 12]
```diff
@@ -219,30 +220,14 @@ class LayoutCalculatorTest {
...
```
Comment (cketti) [inline app/src/test/java/nerd/tuxmobil/fahrplan/congress/schedule/LayoutCalculatorTest.kt:223]: There's no need for a list now. Use `session1` and `session2` directly.

File: database/src/main/java/info/metadude/android/eventfahrplan/database/repositories/RealHighlightsDatabaseRepository.kt [hunk 13]
```diff
@@ -77,6 +78,23 @@ class RealHighlightsDatabaseRepository(
         }
     }
 
+    /**
+     * Resets the value of the [HIGHLIGHT] column to [`false`][HIGHLIGHT_STATE_OFF] for the row
+     * with the given [sessionId].
+     */
+    override fun delete(sessionId: String): Int {
+        with(sqLiteOpenHelper) {
+            return writableDatabase.updateRow(
```
Comment (cketti) [inline database/src/main/java/info/metadude/android/eventfahrplan/database/repositories/RealHighlightsDatabaseRepository.kt:87]: Why update the row and not delete it?

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/models/SessionTest.kt (hunk 10 above)
Comment (johnjohndoe) [reply app/src/test/java/nerd/tuxmobil/fahrplan/congress/models/SessionTest.kt:306]: I can certainly swap these. Is there a particular reason besides readability/consistency?

File: database/src/main/java/info/metadude/android/eventfahrplan/database/repositories/RealHighlightsDatabaseRepository.kt (hunk 13 above)
Comment (johnjohndoe) [reply database/src/main/java/info/metadude/android/eventfahrplan/database/repositories/RealHighlightsDatabaseRepository.kt:87]: Consistency with `deleteAll()` for now. Nothing I can explain right now. I leave the investigation and possible cleanup for a follow-up pull request. I linked the related commit 8bee70b0aa1974e56610cde819f7df04ac8f92e7 in the commit message.

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmServicesTest.kt (hunk 8 above)
Comment (johnjohndoe) [reply app/src/test/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmServicesTest.kt]: :heavy_check_mark: OK. Removed.

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmServicesTest.kt (hunk 9 above)
Comment (johnjohndoe) [reply app/src/test/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmServicesTest.kt]: :heavy_check_mark: OK. Removed.

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/schedule/LayoutCalculatorTest.kt (hunk 11 above)
Comment (johnjohndoe) [reply app/src/test/java/nerd/tuxmobil/fahrplan/congress/schedule/LayoutCalculatorTest.kt]: :heavy_check_mark: Refactored.

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/schedule/LayoutCalculatorTest.kt (hunk 12 above)
Comment (johnjohndoe) [reply app/src/test/java/nerd/tuxmobil/fahrplan/congress/schedule/LayoutCalculatorTest.kt]: :heavy_check_mark: Refactored.

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/models/SessionTest.kt (hunk 10 above)
Comment (cketti) [reply app/src/test/java/nerd/tuxmobil/fahrplan/congress/models/SessionTest.kt]: `session` is an input. But in a test you want to assert that the output meets certain criteria.

**Review by cketti — APPROVED** [review]
